### PR TITLE
Cache Gradle wrapper and dependencies on CI builds

### PR DESCRIPTION
**Pull request by oehme**
## Overview

This should speed up CI builds as Gradle distributions and external dependencies no longer have to be downloaded each time. Since there are also some Maven integration tests, the local Maven repository is also cached.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -101,6 +165,15 @@ jobs:
     if: github.event_name == 'push' && github.repository == 'junit-team/junit5' && (startsWith(github.ref, 'refs/heads/releases/') || github.ref == 'refs/heads/main')
     steps:
     - uses: actions/checkout@v2
+    - name: Cache Gradle wrapper and dependencies
```
Comment (oehme) [inline .github/workflows/main.yml:168]: This step and the one below don't invoke the Maven integration tests, so no need to cache .m2 here.

File: .github/workflows/cross-version.yml [hunk 2]
```diff
@@ -23,6 +23,22 @@ jobs:
     container: "junitteam/build:${{ matrix.jdk }}"
     steps:
       - uses: actions/checkout@v2
+      - name: Cache Gradle wrapper and dependencies
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.gradle/caches/
+            ~/.gradle/wrapper/dists
+          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*') }}
```
Comment (marcphilipp) [inline .github/workflows/cross-version.yml:32]: Shouldn't this also include `buildSrc/**`, `gradle/**`, and `gradle.properties`?

Comment (oehme) [reply .github/workflows/cross-version.yml]: Good point, I've expanded the list.

**Comment by oehme** [comment]
Something's not working corretly here. On Linux, the cache entry is 0MB, on Windows we get no hits at all and on MacOS it works fine.

**Comment by oehme** [comment]
Ah I think I know why - All builds are using the same key, but some of them aren't actually invoking any Gradle or Maven steps. So we have to give each its own ID.

**Comment by oehme** [comment]
Nope, still seeing the same issue. On Windows creating the cache entry fails because of the `.lock` files. On Linux it only works correctly for the "reproducible build" workflow, but not the others. The difference seems to be that the "reproducible build" workflow doesn't use a container, but the others do.

**Comment by sormuras** [comment]
_Who put what into your ☕ , Stefan? On steroids?_ 🚀  ;-)

Watching every commit since yesterday and hoping to see a significant gain of build speed on all platforms. As long as the builds stay "correct" and the configuration of the caching setup isn't too complex, I'm really forward to the (timing) results of this PR.

**Comment by oehme** [comment]
Finally figured out the issue on Linux containers - `~/.gradle`  doesn't work, you have to use `/root/.gradle` instead. The Windows issue was easily fixed by stopping the daemon (releasing its file locks).

You can now compare the performance of a typical CI build on the [main branch](https://github.com/junit-team/junit5/actions/runs/157493860) to [this cached one](https://github.com/junit-team/junit5/actions/runs/159399712) on my branch. Each individual build now takes ~3min instead of ~5min.

If you did the coverage report in parallel to the other tests, the overall build time would come down further.

**Comment by oehme** [comment]
I've optimized this a bit further by only including the essentials in the cache jar, which reduces upload/download times. Saves another ~10s on every build.

**Comment by oehme** [comment]
I tried reducing the size of the cache archives by excluding some files, but that doesn't seem to work. I've opened an [issue](https://github.com/actions/cache/issues/364) for the cache action. If that gets resolved, we could save some more time.

**Comment by sormuras** [comment]
Great work, @oehme!

>  CI  / Mac OS (pull_request) Successful in 14m

What-ever happened here. **14m** is not ... quick.

> If you did the coverage report in parallel to the other tests, the overall build time would come down further.

True. 

> I tried reducing the size of the cache archives by excluding some files, [...]

+1'ed the issue you created for actions/cache ... but what about deleting all `gradle*.zip` files manually before caching?

--

I wonder whether https://github.com/junit-team/junit5/blob/main/src/publishDocumentationSnapshotOnlyIfNecessary.sh should also be integrated directly into the GH workflow configuration file. I created it back then in a dedicated script file to be able to invoke and test it locally as well as on different CI service providers.

**Comment by oehme** [comment]
Yeah I saw those outliers too, but not just on my branch. Hard to tell what went wrong there without some timestamps. Maybe some test using an outside resource that sometimes responds slowly?

We could delete the zip files, but I didn't want to complicate the caching further, it's already a lot of copy-pasta.

**Comment by vlsi** [comment]
> Cache Size: ~633 MB (664085360 B)
> /bin/tar -z -xf /__w/_temp/9da0e536-5cf5-4944-8581-e15877787035/cache.tgz -P -C /__w/junit5/junit5
> Cache restored from key: test-Linux-gradle-2bf12571c9ec30ce7d57c6b569df17c660c961e659cee88c27370df7fc661671

AFAIK there's 5GiB limit: https://github.com/actions/cache#cache-limits, so it can be challenging to fit into the limit.

**Comment by vlsi** [comment]
@oehme , do you think it is faster to download Gradle from the cache rather than from services.gradle.org?
I guess the amount of work should be more-or-less the same: download archive and unpack it.

I see why caching dependencies helps: it reduces the number of Nexus/Artifactory requests, however, gradle-..zip should not gain much from caching. Am I missing something?

**Comment by oehme** [comment]
The main idea is that the caches are co-located, probably in the same datacenter, rather than possibly halfway across the globe.

**Comment by marcphilipp** [comment]
Let's give this a try for a while and see how it evolves over time. Thanks again, @oehme! 👍

**Comment by vlsi** [comment]
I wonder what would be the timings if .gradle/wrapper dir is not cached.

My measurements for pgjdbc sugget that caching wrapper does not improve
build time.

**Comment by marcphilipp** [comment]
But it doesn't make it worse, either, does it?

**Comment by vlsi** [comment]
It makes it worse: when the cache is changed (e.g. dependency is updated),
it increases the cache size, so there's more to be compressed and uploaded

>

**Comment by marcphilipp** [comment]
Thanks for the heads-up. I'll keep an eye on it.

**Comment by sormuras** [comment]
800 megs o' cache

image
